### PR TITLE
yolov7

### PR DESCRIPTION
**Pull request by ACE-777**
Tracking algorithm based on YoloV7 detection method in ```tracking_YOLO_V7```.
All files useful for yolov7 are in directories ```models``` and ```utils```.
The weights are in ```yolov7.pt```.

### CODE REVIEW

File: tracking_YOLO_V7.py [hunk 1]
```diff
@@ -0,0 +1,681 @@
...
+        #     continue
+
+            # get the best size fit
+        ball_size = calc_ball_size(bbox)
+        size_ratio = 1
+        if last_ball_size is not None:
+            size_ratio = ball_size / last_ball_size
+            if size_ratio < 1:
+                size_ratio = 1 / size_ratio
+            # if ball_size > last_ball_size * size_factor or ball_size < last_ball_size / size_factor:
```
Comment (evekeen) [inline tracking_YOLO_V7.py:421]: also why not?

File: tracking_YOLO_V7.py [hunk 2]
```diff
@@ -0,0 +1,681 @@
...
+    im_info = torch.FloatTensor(1)
+    num_boxes = torch.LongTensor(1)
+    gt_boxes = torch.FloatTensor(1)
+
+    im_data = im_data.cuda()
+    im_info = im_info.cuda()
+    num_boxes = num_boxes.cuda()
+    gt_boxes = gt_boxes.cuda()
+
+    # im_data = Variable(im_data, volatile=True)
```
Comment (evekeen) [inline tracking_YOLO_V7.py:469]: please remove commented code

File: tracking_YOLO_V7.py [hunk 3]
```diff
@@ -0,0 +1,681 @@
...
+    if label:
+        tf = max(tl - 1, 1)  # font thickness
+        t_size = cv2.getTextSize(label, 0, fontScale=tl / 3, thickness=tf)[0]
+        c2 = c1[0] + t_size[0], c1[1] - t_size[1] - 3
+        cv2.rectangle(img, c1, c2, color, -1, cv2.LINE_AA)  # filled
+        cv2.putText(img, label, (c1[0], c1[1] - 2), 0, tl / 3, [225, 255, 255], thickness=tf, lineType=cv2.LINE_AA)
+
+
+def xywh2xyxy(x):
+    # Convert nx4 boxes from [x, y, w, h] to [x1, y1, x2, y2] where xy1=top-left, xy2=bottom-right
```
Comment (evekeen) [inline tracking_YOLO_V7.py:184]: is original `x,y` a center? in this case, let's make it clear

File: tracking_YOLO_V7.py [hunk 4]
```diff
@@ -0,0 +1,681 @@
...
+    cv2.rectangle(img, c1, c2, color, thickness=tl, lineType=cv2.LINE_AA)
+    if label:
+        tf = max(tl - 1, 1)  # font thickness
+        t_size = cv2.getTextSize(label, 0, fontScale=tl / 3, thickness=tf)[0]
+        c2 = c1[0] + t_size[0], c1[1] - t_size[1] - 3
+        cv2.rectangle(img, c1, c2, color, -1, cv2.LINE_AA)  # filled
+        cv2.putText(img, label, (c1[0], c1[1] - 2), 0, tl / 3, [225, 255, 255], thickness=tf, lineType=cv2.LINE_AA)
+
+
+def xywh2xyxy(x):
```
Comment (evekeen) [inline tracking_YOLO_V7.py:183]:
```suggestion
def xywh2xyxy(box):
```

File: tracking_YOLO_V7.py [hunk 5]
```diff
@@ -0,0 +1,681 @@
...
+        self.act = nn.SiLU() if act is True else (act if isinstance(act, nn.Module) else nn.Identity())
+
+    def forward(self, x):
+        return self.act(self.bn(self.conv(x)))
+
+    def fuseforward(self, x):
+        return self.act(self.conv(x))
+
+
+class Ensemble(nn.ModuleList):
```
Comment (evekeen) [inline tracking_YOLO_V7.py:51]: so did you copy these classes to delete extra files? I still see these classes in yolov7 files

File: tracking_YOLO_V7.py (hunk 1 above)
Comment (ACE-777) [reply tracking_YOLO_V7.py:421]: Add

File: tracking_YOLO_V7.py (hunk 2 above)
Comment (ACE-777) [reply tracking_YOLO_V7.py:469]: Left only useable for video

**Comment by ACE-777** [comment]
Tracking algorithm based on YoloV7 detection method in ```tracking_YOLO_V7```.
The weights are in ```yolov7.pt```.

File: tracking_YOLO_V7.py (hunk 5 above)
Comment (evekeen) [reply tracking_YOLO_V7.py:362]: still here?

Comment (ACE-777) [reply tracking_YOLO_V7.py:406]: remove experimental.py

**Comment by ACE-777** [comment]
For testing algorihtm use next sceme of directories: 
```
golf-trajectory (git project)
+--prob_data_1
+-----Golf_1
+-----Golf_2
+-----Golf_4
+-----Golf_5
+-----Golf_8
+-----Golf_10
+-----Golf_16
+-----Golf_17
+--prob_annotations
+-----01_001.xml
+----- ...
+--tracking_YOLO_V7.py
+-- ...
```
Algorithm calculates a metric IoU for every image from each directory of prob_data_1. Then, it calculates the average value of the metric IoU for all received results and visualizes in console. For examlpe:
 ```average iou for all warmup videos 54.139223943154015%```.

**Comment by ACE-777** [comment]
Path to dir ```stabil_11_14``` with datasets is defined in a variable ```path_to_datasets_for_warmup```.
To save the original images, create a folder with the name ```save_full_image``` and specify the path in variable ```save_dir_for_full_img```.

The result of the average value of the metric over all datasets, including the first image for each dataset:
--image fragment size for detection 500X500: ```average iou for all warmup videos 13.244572383634485%```;
--image fragment size for detection 600X600: ```average iou for all warmup videos 15.477244443488573%```.

The result of the average value of the metric over all datasets, without the first image for each dataset:
--image fragment size for detection 600X600: ```average iou for all warmup videos 13.435528622377637%```.

File: tracking_YOLO_V7.py [hunk 6]
```diff
@@ -0,0 +1,1155 @@
...
+    video_path = ""  # get video path
+    path_to_datasets_for_warmup = str("C:\\Users\\misha\\golfTrajectory\\stabil_11_14")  # path to dir with datasets
+    dir_of_videos_list = os.listdir(path_to_datasets_for_warmup)
+    for dir_name in range(len(dir_of_videos_list)):
+        if video_path != "":
+            cap = cv2.VideoCapture(video_path)
+            num_images = 0
+            out = None
+        else:
+            imglist = os.listdir(path_to_datasets_for_warmup + r"\\" + dir_of_videos_list[dir_name])  # opt.source
```
Comment (evekeen) [inline tracking_YOLO_V7.py:947]: no backslashes please - use `os.path.join()`

File: tracking_YOLO_V7.py [hunk 7]
```diff
@@ -0,0 +1,1155 @@
...
+        expected_ball_size = 10
+        lost_count = 0
+        last_ball_size = None
+
+        frame_index = 0
+        i = 0
+        num_of_processed_images = 0
+        while (num_of_processed_images < num_images):
+            frame_index = frame_index + 1
+            # if video_path == "":
```
Comment (evekeen) [inline tracking_YOLO_V7.py:983]: remove ALL the commented code, please

File: tracking_YOLO_V7.py [hunk 8]
```diff
@@ -0,0 +1,1155 @@
...
+                im_in = np.array(frame)
+
+                im_in = cv2.flip(im_in, 0)  # flip vertically
+                im_in = cv2.flip(im_in, 1)  # flip horizontally
+
+                if out is None:
+                    fps = cap.get(cv2.CAP_PROP_FPS)
+                    frame_width = int(cap.get(3))
+                    frame_height = int(cap.get(4))
+                    out_path = "Golf Ball/Golf Ball"
```
Comment (evekeen) [inline tracking_YOLO_V7.py:1001]: no need for that, but if you want to, please accept path as an optional cmd argument

File: tracking_YOLO_V7.py [hunk 9]
```diff
@@ -0,0 +1,1155 @@
...
+            det_tic = time.time()
+            """
+            YOLO v7
+            """
+            # test_data, iou from test data
+            save_img = True
+            source, weights, imgsz = 'sub_im.jpg', 'yolov7.pt', 600  # image_size
+
+            # Directories
+            save_dir_for_full_img = "C:\\Users\\misha\\golfTrajectory\\golftrajectory\\save_full_image"
```
Comment (evekeen) [inline tracking_YOLO_V7.py:1049]: cmd argument

File: tracking_YOLO_V7.py [hunk 10]
```diff
@@ -0,0 +1,1155 @@
...
+                lost_count = lost_count + 1
+                pos = predict(tracks)
+                print(f'  lost {lost_count}')
+
+            tracks.append(pos)
+            pos = predict(tracks)
+    for a in range(len(all_iou_metrics_from_warmup)):
+        average_sum += all_iou_metrics_from_warmup[a]
+    average_iou = average_sum/len(all_iou_metrics_from_warmup)
+    print("average iou for all warmup videos {}%".format(average_iou * 100))
```
Comment (evekeen) [inline tracking_YOLO_V7.py:1155]:
```suggestion
    print("average iou for all warmup videos {0:.0f}%".format(average_iou * 100))
```

File: tracking_YOLO_V7.py (hunk 8 above)
Comment (ACE-777) [reply tracking_YOLO_V7.py:1001]: deleted!

File: tracking_YOLO_V7.py (hunk 7 above)
Comment (ACE-777) [reply tracking_YOLO_V7.py:983]: done here!

File: tracking_YOLO_V7.py [hunk 11]
```diff
@@ -0,0 +1,1155 @@
...
+
+                # Inference
+                augment = True
+                with torch.no_grad():  # Calculating gradients would cause a GPU memory leak
+                    pred = model(img, augment=augment)[0]
+
+                # Apply NMS
+                conf_thres = 0.25
+                iou_thres = 0.1
+                classes = 32, 1  # sports ball, bicycle
```
[the remark this answers is not in the document]
Comment (ACE-777) [reply tracking_YOLO_V7.py:1086]: deleted!

File: tracking_YOLO_V7.py (hunk 10 above)
Comment (ACE-777) [reply tracking_YOLO_V7.py:1155]: done!

File: tracking_YOLO_V7.py (hunk 6 above)
Comment (ACE-777) [reply tracking_YOLO_V7.py:947]: done!

File: tracking_YOLO_V7.py [hunk 12]
```diff
@@ -0,0 +1,1164 @@
...
+
+
+def parse_args():
+  """
+  Parse input arguments
+  """
+  parser = argparse.ArgumentParser(description='Tracking algorithm based on yolo v7')
+  parser.add_argument('--dataset', dest='dataset',
+                      help='directory to dir with datasets',
+                      default="C:\\Users\\misha\\golfTrajectory\\stabil_11_14\\stabil_11_14\\")
```
Comment (evekeen) [inline tracking_YOLO_V7.py:920]: let's remove these default paths

File: tracking_YOLO_V7.py [hunk 13]
```diff
@@ -0,0 +1,1164 @@
...
+if __name__ == '__main__':
+    args = parse_args()
+    yolo_classes = np.asarray(['__background__', 'golfball'])
+
+    im_data = torch.FloatTensor(1)
+    im_info = torch.FloatTensor(1)
+    num_boxes = torch.LongTensor(1)
+    gt_boxes = torch.FloatTensor(1)
+
+    im_data = im_data.cuda()
```
Comment (evekeen) [inline tracking_YOLO_V7.py:941]: please, no CUDA for inference

File: tracking_YOLO_V7.py [hunk 14]
```diff
@@ -0,0 +1,1164 @@
...
+                            s += f"{n} {names[int(c)]}{'s' * (n > 1)}, "  # add to string
+
+                        # Write results
+                        for *xyxy, conf, cls in reversed(det):
+                            label = f'{names[int(cls)]} {conf:.2f}'
+                            plot_one_box(xyxy, im0, label=label, color=colors[int(cls)], line_thickness=1)
+                            dets, score = _get_best_det(pred, pos, offset, ratio, last_ball_size, thresh=0.1)  #
+                            if dets != None:  # if detected
+                                pos_bbox_for_full_img = (dets[0], dets[1], dets[2], dets[3])
+                                im_file_save = os.path.join(path_to_datasets_for_warmup, os.path.join(dir_of_videos_list[dir_name],
```
Comment (evekeen) [inline tracking_YOLO_V7.py:1121]: please, don't write into dataset

File: tracking_YOLO_V7.py [hunk 15]
```diff
@@ -0,0 +1,1164 @@
...
+                num_boxes.resize_(1).zero_()
+
+            # pdb.set_trace()
+            det_tic = time.time()
+            """
+            YOLO v7
+            """
+            # test_data, iou from test data
+            save_img = True
+            source, weights, imgsz = 'sub_im.jpg', 'yolov7.pt', 600  # image_size
```
Comment (evekeen) [inline tracking_YOLO_V7.py:1054]:
```suggestion
            source, weights, imgsz = 'sub_im.jpg', 'yolov7.pt', 640  # image_size
```

**Review by evekeen — CHANGES_REQUESTED** [review]
please fix according to comments and apply changes from my PR first

File: tracking_YOLO_V7.py (hunk 12 above)
Comment (ACE-777) [reply tracking_YOLO_V7.py:920]: done!

File: tracking_YOLO_V7.py (hunk 13 above)
Comment (ACE-777) [reply tracking_YOLO_V7.py:941]: remove cuda

File: tracking_YOLO_V7.py (hunk 14 above)
Comment (ACE-777) [reply tracking_YOLO_V7.py:1121]: I write in ```save_full_image```. This dir doesn't connect with dataset:https://github.com/acetrace/golf-trajectory/blob/a8f1825516dbc67d86ba82953bf3c3b3eed07baa/tracking_YOLO_V7.py#L1126

File: tracking_YOLO_V7.py (hunk 15 above)
Comment (ACE-777) [reply tracking_YOLO_V7.py:1054]: Done!

File: tracking_YOLO_V7.py (hunk 14 above)
Comment (evekeen) [reply tracking_YOLO_V7.py]: it has `seq_path` in it, so it must be in the dataset

File: tracking_YOLO_V7.py [hunk 16]
```diff
@@ -0,0 +1,1149 @@
...
+                                s += f"{n} {names[int(c)]}{'s' * (n > 1)}, "  # add to string
+
+                            # Write results
+                            for *xyxy, conf, cls in reversed(det):
+                                label = f'{names[int(cls)]} {conf:.2f}'
+                                plot_one_box(xyxy, im0, label=label, color=colors[int(cls)], line_thickness=1)
+                                dets, score = _get_best_det(pred, pos, offset, ratio, last_ball_size, thresh=0.1)  #
+                                if dets != None:  # if detected
+                                    pos_bbox_for_full_img = (dets[0], dets[1], dets[2], dets[3])
+                                    im_file_save = os.path.join(seq_path, imglist[num_of_processed_images_old])
```
Comment (evekeen) [inline tracking_YOLO_V7.py:1099]: rename `im_file_save` please

**Review by evekeen — CHANGES_REQUESTED** [review]
```
  File "/Users/ivkin/my/golf-trajectory/tracking_YOLO_V7.py", line 1013, in <module>
    im = im_in[:, :, ::-1]  # go throw the end to begin for each number
IndexError: too many indices for array: array is 0-dimensional, but 3 were indexed
```

**Comment by ACE-777** [comment]
> ```
>   File "/Users/ivkin/my/golf-trajectory/tracking_YOLO_V7.py", line 1013, in <module>
>     im = im_in[:, :, ::-1]  # go throw the end to begin for each number
> IndexError: too many indices for array: array is 0-dimensional, but 3 were indexed
> ```
Maybe its connected with im_file from datasets, that are located in that way: ```C:\\Users\\misha\\golfTrajectory\\stabil_11_14\\stabil_11_14\\0ewv4ibe2ce2zn58ti5yu65wzkm2-img_8209-17.27.mov\00000001.jpg```

https://github.com/acetrace/golf-trajectory/blob/4b2768e2fe943d142c7819097729117272113d74/tracking_YOLO_V7.py#L1006-L1007

**Comment by evekeen** [comment]
> > ```
> >   File "/Users/ivkin/my/golf-trajectory/tracking_YOLO_V7.py", line 1013, in <module>
> >     im = im_in[:, :, ::-1]  # go throw the end to begin for each number
> > IndexError: too many indices for array: array is 0-dimensional, but 3 were indexed
> > ```
> 
> Maybe its connected with im_file from datasets, that are located in that way: `C:\\Users\\misha\\golfTrajectory\\stabil_11_14\\stabil_11_14\\0ewv4ibe2ce2zn58ti5yu65wzkm2-img_8209-17.27.mov\00000001.jpg`
> 
> https://github.com/acetrace/golf-trajectory/blob/4b2768e2fe943d142c7819097729117272113d74/tracking_YOLO_V7.py#L1006-L1007

no, it's something else

**Comment by ACE-777** [comment]
Fixed IndexError. Works with a non-img file in the video directory.

**Comment by ACE-777** [comment]
> > > ```
> > >   File "/Users/ivkin/my/golf-trajectory/tracking_YOLO_V7.py", line 1013, in <module>
> > >     im = im_in[:, :, ::-1]  # go throw the end to begin for each number
> > > IndexError: too many indices for array: array is 0-dimensional, but 3 were indexed
> > > ```
> > 
> > 
> > Maybe its connected with im_file from datasets, that are located in that way: `C:\\Users\\misha\\golfTrajectory\\stabil_11_14\\stabil_11_14\\0ewv4ibe2ce2zn58ti5yu65wzkm2-img_8209-17.27.mov\00000001.jpg`
> > https://github.com/acetrace/golf-trajectory/blob/4b2768e2fe943d142c7819097729117272113d74/tracking_YOLO_V7.py#L1006-L1007
> 
> no, it's something else

Fixed IndexError. Works with a non-img file in the video directory.

**Review by evekeen — APPROVED** [review]

**Pull request merged**